store/store.go

```diff
@@ -0,0 +1,142 @@
+package store
+
+import (
+	"encoding/binary"
+	"errors"
+	"sync"
+
+	"go.uber.org/multierr"
+
+	"github.com/lazyledger/optimint/hash"
+	"github.com/lazyledger/optimint/types"
+)
+
+var (
+	blockPrefix  = [1]byte{1}
+	indexPrefix  = [1]byte{2}
+	commitPrefix = [1]byte{3}
+)
+
+type DefaultStore struct {
+	db KVStore
+
+	height uint64
+
+	// mtx protects height
+	mtx sync.RWMutex
+}
+
+var _ Store = &DefaultStore{}
+
+func New() Store {
+	return &DefaultStore{db: NewInMemoryKVStore()}
+}
+
+func (s *DefaultStore) Height() uint64 {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	return s.height
+}
+
+func (s *DefaultStore) SaveBlock(block *types.Block, commit *types.Commit) error {
+	hash, err := hash.Hash(&block.Header)
+	if err != nil {
+		return err
+	}
+	blockKey := getBlockKey(hash)
+	commitKey := getCommitKey(hash)
+	indexKey := getIndexKey(block.Header.Height)
+
+	blockBlob, err := block.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	commitBlob, err := commit.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	// TODO(tzdybal): use transaction for consistency of DB
+	err = multierr.Append(err, s.db.Set(blockKey, blockBlob))
+	err = multierr.Append(err, s.db.Set(commitKey, commitBlob))
+	err = multierr.Append(err, s.db.Set(indexKey, hash[:]))
+
+	if block.Header.Height > s.height {
+		s.height = block.Header.Height
+	}
+
```

Comment:
So we set higher height even if we failed to write to DB? Is it intentional?

Reply:
Good catch, it's not intentional.